### PR TITLE
Hardhat-ethers library linking.

### PR DESCRIPTION
**Pull request by scnale**
This adds an error when creating an `ethers.contractFactory` for an abstract contract and adds an option to link libraries when creating contract factories.

### CODE REVIEW

**Review by fvictorio — APPROVED** [review]

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -35,6 +35,8 @@ jobs:
         run: yarn --frozen-lockfile
       - name: Clean
         run: yarn clean
+      - name: Build
```
Comment (alcuadrado) [inline .github/workflows/CI.yml:38]: Why did you add this? I'm really curious about it.

File: packages/hardhat-ethers/README.md [hunk 2]
```diff
@@ -42,7 +42,16 @@ automatically connected to the selected network.
 These helpers are added to the `ethers` object:
 
 ```typescript
-function getContractFactory(name: string, signer?: ethers.Signer): Promise<ethers.ContractFactory>;
+interface LibraryLinks {
```
Comment (alcuadrado) [inline packages/hardhat-ethers/README.md:45]:
```suggestion
interface Libraries {
```

File: packages/hardhat-ethers/README.md [hunk 3]
```diff
@@ -42,7 +42,16 @@ automatically connected to the selected network.
...
+  [libraryName: string]: string;
+}
+
+interface FactoryOptions {
+  signer?: ethers.Signer;
+  libraryLinks?: LibraryLinks;
```
Comment (alcuadrado) [inline packages/hardhat-ethers/README.md:51]:
```suggestion
  libraries?: Libraries;
```

File: packages/hardhat-ethers/README.md [hunk 4]
```diff
@@ -42,7 +42,16 @@ automatically connected to the selected network.
...
+}
+
+function getContractFactory(name: string, signerOrFactoryOptions?: ethers.Signer | FactoryOptions): Promise<ethers.ContractFactory>;
```
Comment (alcuadrado) [inline packages/hardhat-ethers/README.md:54]: Let's split this into three different signatures here, as this is documentation.

```ts
function getContractFactory(name: string): Promise<ethers.ContractFactory>;

function getContractFactory(name: string, signer: ethers.Signer): Promise<ethers.ContractFactory>;

function getContractFactory(name: string, factoryOptions?: FactoryOptions): Promise<ethers.ContractFactory>;
```

File: packages/hardhat-ethers/README.md [hunk 5]
```diff
@@ -79,6 +88,25 @@ And then run `npx hardhat blockNumber` to try it.
 
 Read the documentation on the [Hardhat Runtime Environment](https://usehardhat.com/advanced/hardhat-runtime-environment.html) to learn how to access the HRE in different ways to use ethers.js from anywhere the HRE is accessible.
 
+### Library linking
+
+Some contracts need to be linked with libraries before they are deployed. You can pass these library links to the `getContractFactory` function with an object like this:
```
Comment (alcuadrado) [inline packages/hardhat-ethers/README.md:93]:
```suggestion
Some contracts need to be linked with libraries before they are deployed. You can pass the addresses of their libraries to the `getContractFactory` function with an object like this:
```

File: packages/hardhat-ethers/README.md [hunk 6]
```diff
@@ -79,6 +88,25 @@ And then run `npx hardhat blockNumber` to try it.
...
+
+```js
+const contractFactory = await this.env.ethers.getContractFactory(
+  "Example",
+  {
+    libraryLinks: {
```
Comment (alcuadrado) [inline packages/hardhat-ethers/README.md:99]:
```suggestion
    libraries: {
```

File: packages/hardhat-ethers/README.md [hunk 7]
```diff
@@ -79,6 +88,25 @@ And then run `npx hardhat blockNumber` to try it.
...
+      ExampleLib: "0x..."
+    }
+  }
+);
+```
+
+This allows you to create a contract factory for the `Example` contract and link its `ExampleLib` library references to the address `"0x..."`.
+
+To create a contract factory, all libraries must be linked. An error will be thrown informing you of any missing library links.
```
Comment (alcuadrado) [inline packages/hardhat-ethers/README.md:108]:
```suggestion
To create a contract factory, all libraries must be linked. An error will be thrown informing you of any missing library.
```

File: packages/hardhat-ethers/README.md (hunk 2 above)
Comment (alcuadrado) [reply packages/hardhat-ethers/README.md:45]: Note that these renames also need to be done in the code itself.

File: packages/hardhat-ethers/src/helpers.ts [hunk 8]
```diff
@@ -24,35 +46,172 @@ export function getContractFactory(
...
       hre,
       nameOrAbi,
-      bytecodeOrSigner as ethers.Signer | undefined
+      bytecodeOrFactoryOptions as ethers.Signer | FactoryOptions | undefined
     );
+
+    if (contractFactory.bytecode === "0x") {
+      throw new NomicLabsHardhatPluginError(
+        pluginName,
+        `The requested contract, ${nameOrAbi}, is an abstract contract.
```
Comment (alcuadrado) [inline packages/hardhat-ethers/src/helpers.ts:64]: `You are trying to create a contract factory for the contract ${nameOrAbi}, which is abstract and can't be deployed.`

Comment (alcuadrado) [reply packages/hardhat-ethers/src/helpers.ts:64]: In general, when a message like this has to be defined, you can discuss it in the `#hardhat-copy` channel.

File: packages/hardhat-ethers/src/helpers.ts [hunk 9]
```diff
@@ -24,35 +46,172 @@ export function getContractFactory(
...
+  }
+
+  const linksToApply: Map<string, Link> = new Map();
+  for (const [linkedLibraryName, linkedLibraryAddress] of Object.entries(
+    libraryLinks
+  )) {
+    if (!ethers.utils.isAddress(linkedLibraryAddress)) {
+      throw new NomicLabsHardhatPluginError(
+        pluginName,
+        `The library name ${linkedLibraryName} has an invalid address: ${linkedLibraryAddress}.`
```
Comment (alcuadrado) [inline packages/hardhat-ethers/src/helpers.ts:128]:
```suggestion
        `You tried to link the contract ${contractName} with the library ${linkedLibraryName}, but provided this invalid address: ${linkedLibraryAddress}`
```

File: packages/hardhat-ethers/src/helpers.ts [hunk 10]
```diff
@@ -24,35 +46,172 @@ export function getContractFactory(
...
+          .map((x) => `* ${x}`)
+          .join("\n");
+        detailedMessage = `The libraries needed are:
+${libraryFQNames}`;
+      } else {
+        detailedMessage = "This contract doesn't need linking any libraries.";
+      }
+      throw new NomicLabsHardhatPluginError(
+        pluginName,
+        `The library name ${linkedLibraryName} does not correspond to any of the libraries needed by the contract ${contractName}.
```
Comment (alcuadrado) [inline packages/hardhat-ethers/src/helpers.ts:153]:
```suggestion
        `You tried to link the contract ${contractName} with ${linkedLibraryName}, which is not one of its libraries.
```

File: .github/workflows/CI.yml (hunk 1 above)
Comment (fvictorio) [reply .github/workflows/CI.yml:38]: Oh, yeah, I forgot to explain this.

The linter was failing in the CI but not locally. After looking into it, we found that a linter rule (the one that forbids you from adding two strings and suggest using interpolation instead) was throwing a false positive because the type inferred for some value was wrong. This only happened if `hardhat-core` wasn't built!

I have no idea why that happened, but the quick solution was to make sure that the build was executed before running the lint :man_shrugging:
